File: 0002. Add Two Numbers/python/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head: Optional[ListNode] = None
        prev: Optional[ListNode] = None
        value = 0

        while l1 or l2 or value > 0:
            value += (l1.val if l1 else 0) + (l2.val if l2 else 0)

            newNode = ListNode(value % 10)
            if prev:
                prev.next = newNode
            else:
                head = newNode
            prev = newNode

            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
            value //= 10

        return head

File: 0002. Add Two Numbers/java/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = null;
        ListNode prev = null;
        int value = 0;

        while (l1 != null || l2 != null || value > 0) {
            value += (l1 != null ? l1.val : 0) + (l2 != null ? l2.val : 0);

            ListNode newNode = new ListNode(value % 10);
            if (prev != null) {
                prev.next = newNode;
            } else {
                head = newNode;
            }
            prev = newNode;

            l1 = (l1 != null ? l1.next : null);
            l2 = (l2 != null ? l2.next : null);
            value /= 10;
        }
        return head;
    }
}
